### PR TITLE
feat(leaveapp): update block leave validation for 4 + half days

### PR DESCRIPTION
**Pull request by Prathmesh-Copods**
Title: Add Half-Day Pattern to Block Leave Validation

1. Current Functionality:
    - Validates block leave with 5 consecutive full days + max 2 WFH days

2. New Addition:
    - Added alternative block leave pattern:
      - 4 consecutive full days + 1 half day
      - Maximum 2 WFH days allowed
      - Half day can be positioned anywhere in the sequence

3. Technical Changes:
    - Updated is_block_leave() to handle both patterns
    - Modified is_block_leave_taken() for consistency
    - Added pattern validation for half-day combinations
    - Maintained existing WFH day limits

4. Testing:
    - Tested with multiple scenarios:
      - Standard block leave pattern (5 full days)
      - New half-day pattern (4 full + 1 half)
      - Various half-day positions in sequence
      - Invalid patterns
      - 90-day restriction validation

No breaking changes to existing functionality.

### CODE REVIEW

File: LeaveTrackingApp/utils.py [hunk 1]
```diff
@@ -681,32 +681,82 @@ def check_leave_overlap(leave_data):
...
+    1. Standard Case: At least 5 consecutive leaves with at most 2 WFH
+    2. Half-day Case: At least 4 consecutive leaves + one half day + at most 2 WFH
+    """
     leave_type_dict = {leave_type.name: str(leave_type.id) for leave_type in LeaveType.objects.filter(name__in=['pto', 'wfh', 'optional_leave'])}
     block_limits = {x.name: int(x.max_days_allowed) for x in RuleSet.objects.filter(name__in=['block_leave_pto', 'block_leave_wfh'])}
-    x = ''.join('1' if ((day['type']==leave_type_dict['pto'] or day['type']==leave_type_dict['optional_leave']) and not day['is_half_day']) else '0' if day['type']==leave_type_dict['wfh'] else 'x' for day in leave_data['day_details'])
-    if x.find("1"*block_limits['block_leave_pto']) < 0 or x.count("0") > block_limits['block_leave_wfh']: 
+    
+    standard_sequence = ''.join('1' if ((day['type']==leave_type_dict['pto'] or day['type']==leave_type_dict['optional_leave']) and not day['is_half_day']) 
+                               else '0' if day['type']==leave_type_dict['wfh']
```
Comment (chandani-copods) [inline LeaveTrackingApp/utils.py:694]: Optional leave should not be in block leave.

File: LeaveTrackingApp/utils.py [hunk 2]
```diff
@@ -681,32 +681,82 @@ def check_leave_overlap(leave_data):
...
+    
+    wfh_count = standard_sequence.count('0')
+    if wfh_count > block_limits['block_leave_wfh']:
         return False
-    return True
+
+    if standard_sequence.find("1"*block_limits['block_leave_pto']) >= 0:
+        return True
+
+    patterns = ['1111H', 'H1111', '11H11', '111H1', '1H111']
```
Comment (chandani-copods) [inline LeaveTrackingApp/utils.py:709]: In block leave, consecutively 4 PTOs should be there atleast.

File: LeaveTrackingApp/utils.py [hunk 3]
```diff
@@ -681,32 +681,82 @@ def check_leave_overlap(leave_data):
...
         Q(leave_type__name='pto') &
         (Q(start_date__lte=end) & Q(end_date__gte=start))
     )
     block_limits = {x.name: int(x.max_days_allowed) for x in RuleSet.objects.filter(name__in=['block_leave_pto', 'block_leave_wfh'])}
+    
     for leave in leaves:
-        x = ''.join('1' if ((day.type.name == 'pto' or day.type.name == 'optional_leave') and not day.is_half_day) else '0' if day.type.name == 'wfh' else 'x' for day in leave.day_details.all().order_by('date'))
-        if x.find("1"*block_limits['block_leave_pto']) >= 0 and x.count("0") <= block_limits['block_leave_wfh']: 
+        standard_sequence = ''.join('1' if ((day.type.name == 'pto' or day.type.name == 'optional_leave') and not day.is_half_day)
+                                  else '0' if day.type.name == 'wfh'
```
Comment (chandani-copods) [inline LeaveTrackingApp/utils.py:736]: Remove optional leave as it can not be the part of block leave.

File: LeaveTrackingApp/utils.py (hunk 1 above)
Comment (Prathmesh-Copods) [reply LeaveTrackingApp/utils.py]: removed Optional leave

File: LeaveTrackingApp/utils.py (hunk 2 above)
Comment (Prathmesh-Copods) [reply LeaveTrackingApp/utils.py]: fixed

File: LeaveTrackingApp/utils.py (hunk 3 above)
Comment (Prathmesh-Copods) [reply LeaveTrackingApp/utils.py:839]: done

**Comment by anuragP04** [comment]
please resolve conflicts

**Review by chandani-copods — APPROVED** [review]

**Pull request merged**